src/main/java/io/pravega/connectors/flink/FlinkPravegaInputFormat.java

```diff
@@ -176,4 +184,36 @@ public T nextRecord(T t) throws IOException {
     public void close() throws IOException {
         this.pravegaReader.close();
     }
+
+    // ------------------------------------------------------------------------
+    //  Utilities
+    // ------------------------------------------------------------------------
+
+    private void setupReaderGroupForCurrentExecution() {
+        try (ReaderGroupManager readerGroupManager = ReaderGroupManager.withScope(scopeName, controllerURI)) {
+            final int currentAttemptNum = getRuntimeContext().getAttemptNumber();
+
+            // delete reader group from last execution attempt, if any
+            if (currentAttemptNum > 0) {
+                try {
+                    readerGroupManager.deleteReaderGroup(getReaderGroupNameForExecutionAttempt(currentAttemptNum - 1));
+                } catch (RuntimeException e) {
+                    // ignore InvalidStreamExceptions, which is the result of duplicate reader group deletions
+                    // (the reader group deletion operation is not idempotent); rethrow all other exceptions
```

Comment:
This part is certainly not that nice.
The main issue is that there currently is no "global failure hook" for input formats when an execution attempt fails, so there is no simple way to delete reader groups only once on failure.

Reply:
I may also be doing this overly complicated.
Is there a better approach to reset the reader group state, instead of creating a new reader group and deleting the previous one?